components/Buttons/src/MDCFloatingButton+Animation.m

```diff
@@ -0,0 +1,242 @@
+    [CATransaction setCompletionBlock:exitActions];
+
+    CABasicAnimation *overallScaleAnimation =
+        [MDCFloatingButton animationWithKeypath:@"transform"
+                                        toValue:[NSValue valueWithCATransform3D:scaleTransform]
+                                      fromValue:nil
+                                 timingFunction:[[CAMediaTimingFunction alloc]
+                                                    initWithControlPoints:0.4f:0.0f:1.0f:1.0f]
+                                       fillMode:kCAFillModeForwards
+                                       duration:kMDCFloatingButtonExitDuration
+                                    beginOffset:0];
+    [self.layer addAnimation:overallScaleAnimation forKey:kMDCFloatingButtonTransformKey];
+
+    CABasicAnimation *iconScaleAnimation =
+        [MDCFloatingButton animationWithKeypath:@"transform"
+                                        toValue:[NSValue valueWithCATransform3D:scaleTransform]
+                                      fromValue:nil
+                                 timingFunction:[[CAMediaTimingFunction alloc]
+                                                    initWithControlPoints:0.0f:0.0f:0.2f:1.0f]
+                                       fillMode:kCAFillModeForwards
+                                       duration:kMDCFloatingButtonExitIconDuration
+                                    beginOffset:kMDCFloatingButtonExitIconOffset];
+    [self.imageView.layer addAnimation:iconScaleAnimation forKey:kMDCFloatingButtonTransformKey];
+
+    CABasicAnimation *opacityAnimation = [MDCFloatingButton
+        animationWithKeypath:@"opacity"
+                     toValue:[NSNumber numberWithFloat:0]
+                   fromValue:nil
+              timingFunction:[CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionLinear]
+                    fillMode:kCAFillModeForwards
+                    duration:kMDCFloatingButtonOpacityDuration
+                 beginOffset:kMDCFloatingButtonOpacityExitOffset];
+    [self.layer addAnimation:opacityAnimation forKey:kMDCFloatingButtonOpacityKey];
+
+    [CATransaction commit];
+  } else {
+    exitActions();
+  }
+}
+
+#pragma mark - UIView frame overrides
+
+/*
+ Override the value of the frame so that it appears to be consistent with how the view would look
+ when it is not scaled. This is necessary because when the view is transformed by an affine scale,
+ the frame will be modified. Apple is very clear that if a UIView's transform is not the identity
+ transform, then the value of the frame is undefined and should be ignored.
+
+ Unfortunately, much of our layout code assumes that frames are valid and relies on them instead of
+ the view's center and bounds for positioning information. As a result, this method will determine
+ what the frame would look like without the transform by using the center and bounds.
+
+ https://developer.apple.com/documentation/uikit/uiview/1622621-frame
+ */
+- (CGRect)frame {
```

Comment:
I think this is too much magic and will burn us at some point in the future. The `frame` property is pretty central and special. For example, the page you link to says

> Changing the frame rectangle automatically redisplays the receiver without invoking the draw(_:) method.

How would that interact with this? Are we sure that this will continue to work in the future? 

Let's chat and figure out a more maintainable solution.